core/distributed/vector.cpp

```diff
@@ -0,0 +1,389 @@
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#include "core/distributed/vector_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace vector {
+namespace {
+
+
+GKO_REGISTER_OPERATION(compute_norm2_sqr, dense::compute_norm2_sqr);
+GKO_REGISTER_OPERATION(compute_sqrt, dense::compute_sqrt);
+GKO_REGISTER_OPERATION(outplace_absolute_dense, dense::outplace_absolute_dense);
+GKO_REGISTER_OPERATION(build_local, distributed_vector::build_local);
+
+
+}  // namespace
+}  // namespace vector
+
+
+namespace detail {
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void read_distributed_impl(
+    const device_matrix_data<ValueType, GlobalIndexType>& data,
+    const Partition<LocalIndexType, GlobalIndexType>* partition,
+    Vector<ValueType>* result)
+{
+    auto exec = result->get_executor();
+
+    auto rank = result->get_communicator().rank();
+    result->get_local()->fill(zero<ValueType>());
+    exec->run(vector::make_build_local(
+        data, make_temporary_clone(exec, partition).get(), rank,
+        result->get_local()));
+}
+
+#define GKO_DECLARE_DISTRIBUTED_READ_DISTRIBUTED_IMPL(               \
+    ValueType, LocalIndexType, GlobalIndexType)                      \
+    void read_distributed_impl(                                      \
+        const device_matrix_data<ValueType, GlobalIndexType>& data,  \
+        const Partition<LocalIndexType, GlobalIndexType>* partition, \
+        Vector<ValueType>* storage)
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_LOCAL_GLOBAL_INDEX_TYPE(
+    GKO_DECLARE_DISTRIBUTED_READ_DISTRIBUTED_IMPL);
+
+
+}  // namespace detail
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* alpha, const LinOp* b,
+                                   const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size)
+    : Vector(exec, comm, global_size, local_size, local_size[1])
+{}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size, size_type stride)
+    : EnableLinOp<Vector<ValueType>>{exec, global_size},
+      DistributedBase{comm},
+      local_{exec, local_size, stride}
+{}
```

Comment:
We should probably check that the number of rhs matches for global size and local size?